BUOY: refactor and fix mission and perception (backup plan)

kev-the-dev opened the pull request:
* refactor much of the buoy finder python node. I found many bugs and added documentation. I also added support for the LAB color space, and set new thresholds in LAB based on the bags I could find. Though it still just uses color to segment, it seems to work pretty well now is various lightning conditions, but a better approach would be easy to drop in.
* rewrite the Buoy mission. It makes better use of txros now and should solve the mission faster in some cases (for example, it asynchronously stops searching as soon as it has a pose on a buoy, then continues searching after it is bumped)
* Fix a problem in sim where odom start at [0, 0, 0], meaning that some underwater moves will be canceled because they are positive z. Odom in sim now starts at [0, 0, initial z]. 
* some other small bugs in vision tools and launch files have been fixed

kev-the-dev committed:
BUOY: move buoy_2d.py to buoy_finder.py

kev-the-dev committed:
VISION TOOLS: allow custom topic for rviz publisher

kev-the-dev committed:
VISION TOOLS: remove debug spam on multi_observation class

kev-the-dev committed:
BUOY: refactor and fix bugs in buoy_finder node

* add support for LAB colorspace for segmentaiton

* fix a number of bugs found in testing

* add comments in many undocumented functions

* set new threshold values based on LAB

kev-the-dev committed:
SIMULATION: make odometry Z correct depth in gazebo

kev-the-dev committed:
BUOY: edge case bug with service callback

kev-the-dev committed:
MISSIONS: refactor and fix bugs in buoy mission

kev-the-dev commented on command/sub8_missions/missions/buoy.py:
```diff
@@ -2,132 +2,98 @@
+            for color in self.found:
+                if self.found[color] is not None and color not in self.bumped:
+                    pattern.cancel()
+                    yield self.bump(color)
```
So right now it just bumps buoys as it finds them. This will actually not maximize points as order matters. I'll work on this.

kev-the-dev replied:
Did it!

kev-the-dev committed:
FORMAT: fix pep8 errors in PR#275

kev-the-dev committed:
BUOY: bump in correct order when possible

kev-the-dev committed:
BUOY: make mission search radius a constant

mattlangford commented on perception/sub8_perception/nodes/buoy_finder.py:
```diff
@@ -0,0 +1,364 @@
+                               np.array(rospy.get_param(high))]
+            rospy.loginfo("BUOY - Thresholds for {} buoy loaded using {} colorspace".format(
+                          self.color, self.color_space))
+            return
```
Should this `return` be here? The debug stuff won't work unless there are no color space params.

kev-the-dev replied:
True.

mattlangford commented on perception/sub8_perception/nodes/buoy_finder.py:
```diff
@@ -0,0 +1,364 @@
+                cv2.createTrackbar('high {}'.format(self.color_space[i]), self.color, int(self.thresholds[1][i]), 255,
+                                   lambda x, _ind=i: change_threshold(1, _ind, x))
+
+    def segment(self, img):
```
This isn't really a segment function, more like `get_mask`

mattlangford commented on perception/sub8_perception/nodes/buoy_finder.py:
```diff
@@ -0,0 +1,364 @@
+            buoy.pose_pairs.append((t, R))
+
+        if len(buoy.observations) > self.min_observations:
+            self.buoys[buoy_type].est = self.multi_obs.lst_sqr_intersection(buoy.observations, buoy.pose_pairs)
```
I still don't think this a proper way to do this. The buoys do move under water. Last year they had a buoy attached to something on the surface and that one moved around like crazy. Even with truncating the length of the observations at 300, you'll keep 20 seconds or so of history. From my experience last year, the buoy could move a decent amount in that time.

kev-the-dev replied:
Yeah... maybe expect n observations in x minutes, than clear the buffer if it hasn't been observed in x minuets?

mattlangford reviewed (commented):
This is the backup plan? What's the primary plan?

kev-the-dev commented:
This is really just getting the old threshold based buoy stuff working again. Ideally we will be using @aaron-iglesias's machine learning solution for segmentation if it performs well.

kev-the-dev committed:
BUOY: throw out old observations, PR comments

kev-the-dev committed:
BUG: fix gtk error on importing sub8_vision_tools

* attempted to open gtk windows without being promted, causing problems over ssh

kev-the-dev committed:
BUOY: fix bugs found at test day

BUOY: configure blind moves easily in mission

BUOY: use odom insted of tf for sanity velocity check

BUOY: widen thresholds a bit (sigh)

DSsoto commented on perception/sub8_perception/sub8_vision_tools/rviz.py:
```diff
@@ -14,8 +14,8 @@ class RvizVisualizer(object):
 
-    def __init__(self):
-        self.rviz_pub = rospy.Publisher("visualization/projection", visualization_msgs.Marker, queue_size=3)
+    def __init__(self, topic="visualization/projection"):
```
Can you use a more specific topic name here?

kev-the-dev replied:
This is just the default. Individual nodes (like buoy_finder) will override it. I think we should go through and get all perception nodes publishing to the same rviz marker topic later

DSsoto replied:
I disagree about publishing all markers to the same topic, unless everyone publishes to both a specific and a general topic. It's often very useful to reduce clutter as much as possible when visualizing.

This is ok though.

DSsoto reviewed (approved):
Looks good, did you catch anything that needed to be fixed last pool day?

kev-the-dev commented:
Just pushed the little fixes from test day. Still runs well in sim.
